// mui-quill/dist/RichInputBase/styles/quill.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("quill/dist/quill.core.css");
var styles_1 = require("@material-ui/core/styles");
/**
 * Quill Styles
 */
var useQuillStyles = styles_1.makeStyles(function (theme) {
    return styles_1.createStyles({
        '@global': {
            '.ql-container': {
                position: 'relative',
                width: '100%',
            },
            '.ql-clipboard': {
                left: -100000,
                height: 1,
                overflowY: 'hidden',
                position: 'absolute',
                top: '50%',
            },
            '.ql-editor': {
                padding: 0,
                outline: 'none',
            },
            '.ql-editor p': {
                margin: 0,
                padding: 0,
            },
            '.ql-editor.ql-blank::before': {
                color: 'currentColor',
                opacity: theme.palette.type === 'light' ? 0.42 : 0.5,
                transition: theme.transitions.create('opacity', {
                    duration: theme.transitions.duration.shorter,
                }),
                content: 'attr(data-placeholder)',
                pointerEvents: 'none',
                position: 'absolute',
            },
            '.ql-divider': {
                margin: theme.spacing(0.5, 0.5),
                height: "calc(100% - " + theme.spacing(1) + "px)",
            },
            '.ql-active': {
                backgroundColor: styles_1.fade(theme.palette.grey[400], 0.25) + " !important",
            },
        },
    });
});
// DEFAULT EXPORT
exports.default = useQuillStyles;
